src/osd/ExtentCache.h

```diff
@@ -377,16 +397,37 @@ class ExtentCache {
       tid = in_tid;
       pin_type = in_type;
     }
+    virtual void pin(extent& ext) {
+      pin_list.push_back(ext);
+    }
   };
 
   void release_pin(pin_state &p) {
     for (auto iter = p.pin_list.begin(); iter != p.pin_list.end(); ) {
-      unique_ptr<extent> extent(&*iter); // we now own this
-      iter++; // unlink will invalidate
-      assert(extent->parent_extent_set);
-      auto &eset = *(extent->parent_extent_set);
-      extent->unlink();
-      remove_and_destroy_if_empty(eset);
+      assert(iter->parent_extent_set);
+      if (iter->recommend_reserve) {
+	entity_name_t from = iter->from;
+	shared_ptr<reserve_pin> rpin;	
+
+	if(!reserve_pin_list.find(from, rpin)) {
+	  rpin = shared_ptr<reserve_pin>(new reserve_pin(this));
+	  open_reserve_pin(*rpin);
+	  reserve_pin_list.add(from, rpin);
```

Comment:
I don't think we want to create a new pin for each extent -- particularly not one wrapped in a shared_ptr.  The goal here is to minimize indirection.  I think all you need is a list ordered by recency for eviction?  Maybe a set of those lists per client?

Reply:
Even better if lru_map doesn't move or copy internally -- you could directly embed reserve_pin.